File: datagate/infra/shared/jobs/experiments/pyspark/tlc_trip_record/clean_data.py
import argparse
from dataclasses import dataclass
from time import perf_counter

from pyspark.sql import DataFrame, SparkSession
from pyspark.sql.functions import col, lit, to_timestamp, trim


JOB_NAME = "clean_data"


@dataclass
class JobConfig:
    source_table: str
    target_table: str
    processing_date_hour: str


def parse_args() -> JobConfig:
    parser = argparse.ArgumentParser(
        description="Clean one date_hour from bronze Iceberg table and write to silver Iceberg table"
    )

    parser.add_argument(
        "--source_table",
        required=False,
        default="bronze.yellow_tripdata",
    )

    parser.add_argument(
        "--target_table",
        required=False,
        default="silver.cleaned_yellow_tripdata",
    )

    parser.add_argument(
        "--processing_date_hour",
        required=True,
        help="Processing batch partition, format: yyyy-MM-dd HH:mm:ss",
    )

    args = parser.parse_args()

    return JobConfig(
        source_table=args.source_table,
        target_table=args.target_table,
        processing_date_hour=args.processing_date_hour,
    )


def build_spark_session() -> SparkSession:
    return (
        SparkSession.builder
        .appName(JOB_NAME)
        .getOrCreate()
    )


def read_from_iceberg(
    spark: SparkSession,
    source_table: str,
    processing_date_hour: str,
) -> DataFrame:
    return (
        spark.read
        .format("iceberg")
        .load(f"iceberg.{source_table}")
        .filter(col("processing_date_hour") == to_timestamp(lit(processing_date_hour)))
    )


def write_to_iceberg(
    df: DataFrame,
    target_table: str,
) -> None:
    (
        df.writeTo(f"iceberg.{target_table}")
        .overwritePartitions()
    )


def clean_data(df: DataFrame) -> DataFrame:
    return (
        df
        # Required fields
        .filter(col("vendor_id").isNotNull())
        .filter(col("tpep_pickup_datetime").isNotNull())
        .filter(col("tpep_dropoff_datetime").isNotNull())
        .filter(col("trip_distance").isNotNull())
        .filter(col("fare_amount").isNotNull())
        .filter(col("total_amount").isNotNull())
        .filter(col("date_hour").isNotNull())

        # Trip time validation
        .filter(col("tpep_dropoff_datetime") >= col("tpep_pickup_datetime"))

        # Passenger count validation
        .filter(col("passenger_count").isNull() | (col("passenger_count") >= 0))
        .filter(col("passenger_count").isNull() | (col("passenger_count") <= 8))

        # Trip distance validation
        .filter(col("trip_distance").isNull() | (col("trip_distance") > 0))

        # Location validation
        .filter(col("pulocationid").isNull() | (col("pulocationid") > 0))
        .filter(col("dolocationid").isNull() | (col("dolocationid") > 0))

        # Amount validation
        .filter(col("fare_amount").isNull() | (col("fare_amount") >= 0))
        
        .filter(col("extra").isNull() | (col("extra") >= 0))
        .filter(col("mta_tax").isNull() | (col("mta_tax") >= 0))
        .filter(col("tip_amount").isNull() | (col("tip_amount") >= 0))
        .filter(col("tolls_amount").isNull() | (col("tolls_amount") >= 0))
        .filter(col("improvement_surcharge").isNull() | (col("improvement_surcharge") >= 0))
        .filter(col("total_amount").isNull() | (col("total_amount") >= 0))
        .filter(col("congestion_surcharge").isNull() | (col("congestion_surcharge") >= 0))
        .filter(col("airport_fee").isNull() | (col("airport_fee") >= 0))
        .filter(col("cbd_congestion_fee").isNull() | (col("cbd_congestion_fee") >= 0))

        # Normalize string flag, keep the same schema
        .withColumn("store_and_fwd_flag", trim(col("store_and_fwd_flag")))
        .withColumn("passenger_count", col("passenger_count").cast("bigint"))
        .withColumn("ratecode_id", col("ratecode_id").cast("bigint"))

        # Remove fully duplicated rows
        .dropDuplicates()
    )


def attach_processing_date_hour(df: DataFrame, processing_date_hour: str) -> DataFrame:
    return df.withColumn(
        "processing_date_hour",
        to_timestamp(lit(processing_date_hour)),
    )


def main() -> None:
    job_start_time = perf_counter()

    job_config = parse_args()
    spark = build_spark_session()

    try:
        print(
            f"[{JOB_NAME}] Loading bronze partition "
            f"processing_date_hour={job_config.processing_date_hour}"
        )

        bronze_df = read_from_iceberg(
            spark=spark,
            source_table=job_config.source_table,
            processing_date_hour=job_config.processing_date_hour,
        )

        cleaned_df = attach_processing_date_hour(
            df=clean_data(bronze_df),
            processing_date_hour=job_config.processing_date_hour,
        )

        write_to_iceberg(
            df=cleaned_df,
            target_table=job_config.target_table,
        )

        total_seconds = perf_counter() - job_start_time
        print(f"[Job Completed] {JOB_NAME} finished in {total_seconds:.3f} seconds")

    finally:
        spark.stop()


if __name__ == "__main__":
    main()

File: experiments/experiment_jobs/tlc_trip_record_etl/src/main/scala/clean_data.scala
import org.apache.spark.sql.{DataFrame, SparkSession}
import org.apache.spark.sql.functions._

import java.time.{Duration, Instant}

object clean_data {

  case class JobConfig(
                        source_table: String,
                        target_table: String
                      )

  private val job_name = "clean_data"

  def parse_args(args: Array[String]): JobConfig = {
    JobConfig(
      source_table = get_optional_arg(args, "source_table", "bronze.yellow_tripdata"),
      target_table = get_optional_arg(args, "target_table", "silver.cleaned_yellow_tripdata")
    )
  }

  private def get_optional_arg(args: Array[String], name: String, default_value: String): String = {
    val index = args.indexOf(s"--${name}")

    if (index == -1 || index + 1 >= args.length) {
      default_value
    } else {
      args(index + 1)
    }
  }

  private def build_spark_session(): SparkSession = {
    SparkSession.builder()
      .appName(job_name)
      .getOrCreate()
  }

  private def read_from_iceberg(spark: SparkSession, source_table: String): DataFrame = {
    spark.read
      .format("iceberg")
      .load(s"iceberg.${source_table}")
  }

  private def write_to_iceberg(df: DataFrame, target_table: String): Unit = {
    df.write
      .format("iceberg")
      .mode("append")
      .save(s"iceberg.${target_table}")
  }

  private def clean_data(df: DataFrame): DataFrame = {
    df
      // Loại bỏ các dòng thiếu dữ liệu bắt buộc
      .filter(col("vendorid").isNotNull)
      .filter(col("tpep_pickup_datetime").isNotNull)
      .filter(col("tpep_dropoff_datetime").isNotNull)
      .filter(col("trip_distance").isNotNull)
      .filter(col("fare_amount").isNotNull)
      .filter(col("total_amount").isNotNull)
      .filter(col("date_hour").isNotNull)

      // Làm sạch thời gian chuyến đi
      .filter(col("tpep_dropoff_datetime") >= col("tpep_pickup_datetime"))

      // Làm sạch số lượng hành khách
      .filter(col("passenger_count").isNull || col("passenger_count") >= 0)
      .filter(col("passenger_count").isNull || col("passenger_count") <= 8)

      // Làm sạch khoảng cách chuyến đi
      .filter(col("trip_distance") > 0)

      // Làm sạch mã location
      .filter(col("pulocationid").isNull || col("pulocationid") > 0)
      .filter(col("dolocationid").isNull || col("dolocationid") > 0)

      // Làm sạch các trường tiền
      .filter(col("fare_amount") >= 0)
      .filter(col("extra").isNull || col("extra") >= 0)
      .filter(col("mta_tax").isNull || col("mta_tax") >= 0)
      .filter(col("tip_amount").isNull || col("tip_amount") >= 0)
      .filter(col("tolls_amount").isNull || col("tolls_amount") >= 0)
      .filter(col("improvement_surcharge").isNull || col("improvement_surcharge") >= 0)
      .filter(col("total_amount") >= 0)
      .filter(col("congestion_surcharge").isNull || col("congestion_surcharge") >= 0)
      .filter(col("airport_fee").isNull || col("airport_fee") >= 0)
      .filter(col("cbd_congestion_fee").isNull || col("cbd_congestion_fee") >= 0)

      // Chuẩn hóa string flag nhưng vẫn giữ nguyên cột cũ
      .withColumn("store_and_fwd_flag", trim(col("store_and_fwd_flag")))

      // Loại bỏ dòng trùng hoàn toàn
      .dropDuplicates()
  }

  private def start_timer(step_name: String): Instant = {
    val start_time = Instant.now()
    println(s"[Performance] START - ${step_name} at ${start_time}")
    start_time
  }

  private def end_timer(step_name: String, start_time: Instant): Unit = {
    val end_time = Instant.now()
    val duration_ms = Duration.between(start_time, end_time).toMillis
    val duration_seconds = duration_ms / 1000.0

    println(s"[Performance] END   - ${step_name} at ${end_time}")
    println(s"[Performance] TIME  - ${step_name}: ${duration_ms} ms (${duration_seconds} seconds)")
  }

  private def log_job_config(job_config: JobConfig): Unit = {
    println("[Clean Job] Starting clean data job")
    println(s"Source table: ${job_config.source_table}")
    println(s"Target table: ${job_config.target_table}")
  }

  def main(args: Array[String]): Unit = {
    val job_start_time = Instant.now()
    println(s"[Performance] JOB START at ${job_start_time}")

    val parse_start_time = start_timer("Parse arguments")
    val job_config = parse_args(args)
    end_timer("Parse arguments", parse_start_time)

    val spark_start_time = start_timer("Build Spark session")
    val spark = build_spark_session()
    end_timer("Build Spark session", spark_start_time)

    try {
      log_job_config(job_config)

      val read_start_time = start_timer("Read bronze Iceberg table")
      val bronze_df = read_from_iceberg(
        spark = spark,
        source_table = job_config.source_table
      )
      end_timer("Read bronze Iceberg table", read_start_time)

      val clean_start_time = start_timer("Clean bronze data")
      val cleaned_df = clean_data(bronze_df)
      end_timer("Clean bronze data", clean_start_time)

      val write_start_time = start_timer("Write cleaned data to silver Iceberg table")
      write_to_iceberg(
        df = cleaned_df,
        target_table = job_config.target_table
      )
      end_timer("Write cleaned data to silver Iceberg table", write_start_time)

      val job_end_time = Instant.now()
      val total_duration_ms = Duration.between(job_start_time, job_end_time).toMillis
      val total_duration_seconds = total_duration_ms / 1000.0

      println(s"[Performance] JOB END at ${job_end_time}")
      println(s"[Performance] TOTAL TIME: ${total_duration_ms} ms (${total_duration_seconds} seconds)")
      println("[Clean Job] Clean data job completed successfully")

    } finally {
      val stop_start_time = start_timer("Stop Spark session")
      spark.stop()
      end_timer("Stop Spark session", stop_start_time)
    }
  }
}